### ISSUE TITLE
sp_BlitzWho getting gigantic memory grants on servers with lots of RAM

### ISSUE THREAD

**Issue by AratBrat** [BrentOzarULTD/SQL-Server-First-Responder-Kit#2238]
**Version of the script**
7.91 - December 2019

**What is the current behavior?**
sp_BlitzCache is showing that a particular query in sp_BlitzWho is getting a 346GB memory grant but only using 800KB of it. This is on a 2TB server that has a lot of activity.

Query is attached. When executed by sp_BlitzWho via dynamic SQL, it occurs right after the INSERT @blocked/SELECT query. But in sp_BlitzWho it's somewhere else. The start of building the query occurs here:     /* Think of the StringToExecute as starting with this, but we'll set this up later depending on whether we're doing an insert or a select:
    SELECT @StringToExecute = N'SELECT  GETDATE() AS run_date ,
    */

[BlitzWho_LargeMemoryGrantQuery.txt](https://github.com/BrentOzarULTD/SQL-Server-First-Responder-Kit/files/4037059/BlitzWho_LargeMemoryGrantQuery.txt)

![BlitzWho_LargeMemoryGrantQuery_HolyRows](https://user-images.githubusercontent.com/59666823/72013441-b377e280-3212-11ea-8a23-169a715471ef.png)

![BlitzWho_LargeMemoryGrantQuery_MemoryGrantInfo](https://user-images.githubusercontent.com/59666823/72013454-b8d52d00-3212-11ea-94a2-73e87bf3f40b.png)

![BlitzWho_LargeMemoryGrantQuery_BlitzCacheMemoryGrant](https://user-images.githubusercontent.com/59666823/72013459-bbd01d80-3212-11ea-8194-6ba805b7efdb.png)

**If the current behavior is a bug, please provide the steps to reproduce.**
Run sp_BlitzWho on a server with a lot of RAM and presumably a lot of activity.

**What is the expected behavior?**
A much smaller memory grant. 

I think this can be done by checking @ProductVersionMajor and @ProductVersionMinor and adding OPTION(MAX_GRANT_PERCENT) if the feature is available.

**Which versions of SQL Server and which OS are affected by this issue? Did this work in previous versions of our procedures?**
This is on SQL Server 2016, but I'm imagine it's happening on other servers with similar characteristics (large amount of RAM, heavy activity).

**Comment by BrentOzar** [comment]
A PERFECT EXCUSE TO BUY 1.5TB OF RAM FOR MY MAC PRO, WOOHOO

**Comment by AratBrat** [comment]
@erikdarlingdata helped with the specs of this new server. Maybe he could help you too. ;-)

**Comment by erikdarlingdata** [comment]
Yeah, I had an issue for this a while back, but closed it when it got deprecated.

https://github.com/BrentOzarULTD/SQL-Server-First-Responder-Kit/issues/2066
________________________________

**Comment by BrentOzar** [comment]
OK, added in dev. Just making numbers up - if the server has >50GB target memory, then I added a max grant hint of 5%. Can you try that out and see how that goes? I'm okay with changing both of those numbers (50GB & 5%) or getting fancier if y'all want.

**Comment by AratBrat** [comment]
The service here is spectacular! 

Just tested on the 2TB server. Gets an 18GB memory grant now. Much better. It's now #2 in the BlitzCache memory grant list instead of #1 and a whole digit longer than all the others.

**Comment by erikdarlingdata** [comment]
I would drop it down to 1 or less tbh.
________________________________

**Comment by BrentOzar** [comment]
1% it is! Also, threw in a recompile hint in there to foil Tara's monitoring, HA HA HO HO!

**Comment by erikdarlingdata** [comment]
I THINK IT'S WITHIN SCOPE TO REWRITE IT FROM SCRATCH
________________________________

**Comment by BrentOzar** [comment]
The least I can do is put a cursor in there while it's open, amirite?

**Comment by MyKarpenko** [comment]
@BrentOzar   if there is a chance to add the same hint 
OPTION (MAX_GRANT_PERCENT = 1, RECOMPILE)
instead of OPTION ( RECOMPILE)

to sp_BlitzFirst it should be more stable at the moment a server is under RESOURCE_SEMAPHORE and the proc is only anbe to show the first sp_BlitsWho portion

up to you
I've just learnt today that sp_BlitzWho was tuned with OPTION (MAX_GRANT_PERCENT = 1, RECOMPILE)

**Comment by erikdarlingdata** [comment]
@MyKarpenko:

There are two separate issues here:
1. sp_BlitzWho asks for an unnecessarily large memory grant because of the presentation `order by` on output that includes query text and two query plan columns (one normal, one live) 
2. sp_BlitzFirst gets stuck on the analysis queries in many places (both since startup and sampled) when it's blocked or the queries just perform poorly

I've seen it a lot with the analysis queries that hit tempdb and look for long running/blocking transactions. Adding the memory grant hint to those won't do much to alleviate their stuckness.

- If there are specific queries in sp_BlitzFirst that you see getting stuck on RESOURCE_SEMAPHORE, please open an issue that documents those queries where the hint would be effective
- If you see the queries getting stuck for other reasons, please note if they're blocked by something, or if they're just performing poorly for other reasons (grabbing the query plans is important for this)

It would not be useful to sprinkle max grant percent hints on every query in the stored procedure _just in case_.

I know it sounds a bit double-fisted, but running [sp_WhoIsActive](https://github.com/amachanic/sp_whoisactive/blob/master/sp_WhoIsActive.sql) while sp_BlitzFirst is stuck can help troubleshoot where things are going wrong and why.

`sp_WhoIsActive @get_plans = 1;`

Thanks!

**Comment by MyKarpenko** [comment]
Makes sense
I will gather the details
